samza-core/src/main/java/org/apache/samza/operators/impl/WindowOperatorImpl.java

```diff
@@ -18,26 +18,323 @@
+import org.slf4j.LoggerFactory;
 
-public class WindowOperatorImpl<M, WK, WV> extends OperatorImpl<M, WindowPane<WK, WV>> {
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.PriorityQueue;
+import java.util.function.BiFunction;
+import java.util.function.Function;
 
-  private final WindowInternal<M, WK, WV> window;
+/**
+ * Implementation of a window operator that groups messages into finite windows for processing.
+ *
+ * <p>Notes:
+ * This class implements the processing logic for various types of windows and triggers. It tracks and manages state for
+ * all open windows, the active triggers that correspond to each of the windows and the pending callbacks. It provides
+ * an implementation of {@link TriggerContext} that {@link TriggerImpl}s can use to schedule and cancel callbacks. It
+ * also orchestrates the flow of messages through the various {@link TriggerImpl}s.
+ *
+ * @param <M>  the type of the incoming {@link MessageEnvelope}
+ * @param <K>  the type of the key in this {@link org.apache.samza.operators.MessageStream}
+ * @param <WK> the type of the key in the emitted window pane
+ * @param <WV> the type of the value in the emitted window pane
+ * @param <WM> the type of the emitted window pane
+ *
+ * TODO: Implement expiration of entries and triggers
+ */
+public class WindowOperatorImpl<M extends MessageEnvelope, K, WK, WV, WM extends WindowPane<K, WV>> extends OperatorImpl<M, WM> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(WindowOperatorImpl.class);
 
-  public WindowOperatorImpl(WindowOperatorSpec spec, MessageStreamImpl<M> source, Config config, TaskContext context) {
-    // source, config, and context are used to initialize the window kv-store
+  private final PriorityQueue<TriggerTimerState> pendingCallbacks = new PriorityQueue<>();
+  private final WindowInternal<M, K, WV> window;
+  private final KeyValueStore<WindowKey<K>, WV> store = new InternalInMemoryStore<>();
+
+  private final Map<TriggerKey<K>, TriggerImplWrapper> triggers = new HashMap<>();
+
+  private enum TriggerType { EARLY, DEFAULT, LATE }
+
+  public WindowOperatorImpl(WindowOperatorSpec<M, WK, WV> spec) {
     window = spec.getWindow();
   }
 
   @Override
   public void onNext(M message, MessageCollector collector, TaskCoordinator coordinator) {
+    WindowKey<K> storeKey =  getStoreKey(message);
+    BiFunction<M, WV, WV> foldFunction = window.getFoldFunction();
+    WV wv = store.get(storeKey);
+
+    if (wv == null) {
+      wv = window.getInitializer().get();
+    }
+
+    WV newVal = foldFunction.apply(message, wv);
+    store.put(storeKey, newVal);
+
+    if (window.getEarlyTrigger() != null) {
+      TriggerKey<K> triggerKey = new TriggerKey<>(TriggerType.EARLY, storeKey);
+      getOrCreateTriggerWrapper(triggerKey, window.getEarlyTrigger())
+          .onMessage(message, collector, coordinator);
+    }
+
+    if (window.getDefaultTrigger() != null) {
```

Comment:
Good point, processing time windows do not have late arrivals. All emissions are either early or on time.